### ISSUE TITLE
Tree focus/selection collection holds on to old model elements, after splice call

### ISSUE THREAD

**Issue by roblourens** [Microsoft/vscode#67768]
I'm confused about where the IdentityProvider is used and where object identity is used in the new tree. For example in https://github.com/Microsoft/vscode/commit/98e5b5c0a30591571b5a39aa921cc3f27cea27c9 the tree uses the identity provider to determine whether the focused item was removed, but the model actually keeps a map of elements and looks up items by their object identity. https://github.com/Microsoft/vscode/blob/c3b9299c9d68bb566d946e8f7260177471be8de3/src/vs/base/browser/ui/tree/objectTreeModel.ts#L181

The settings trees have had issues with this because I rebuild the whole settings model so I am setting children with the same IDs that are new objects. So what I I ran into while fixing https://github.com/Microsoft/vscode/issues/67757 is that if I focus some element, then call setChildren with a new model, then essentially call `this.tocTree.setSelection(this.tocTree.getFocus())`, that would fail. The exact element returned by getFocus is no longer in the tree, because it was replaced by a new object with the same ID.

It seems like if elements have IDs then any two elements with the same ID should be equivalent? Or should the settings tree not count on this?

**Comment by joaomoreno** [comment]
The identity provider is purely used for the purposes of preserving focus/selection when replacing elements in the tree. Model elements which share the same identity should keep staying focused/selected across tree splices even when the model element itself changes.

https://github.com/Microsoft/vscode/issues/67890 fixed the `this.tocTree.setSelection(this.tocTree.getFocus())` issue. The traits weren't properly updated across splices (`updateChildren` calls).

Does that help?

**Comment by roblourens** [comment]
I still have the same issue. If I understand that change correctly, it won't help here because it doesn't replace the actual element objects in the Trait's nodes array. It considers the identity provider when updating them which is right, but when I take that element and call setSelection with it, it can still fail because it's an old object. I'll push a branch that demonstrates the issue.

**Comment by roblourens** [comment]
- Check out `roblou/invalidTreeFocus`
- Open the settings editor
- Reload the window
- See error that setSelection failed

Basically the tree fires onDidChangeFocus, the handler waits on setTimeout (this is working around another issue but it shouldn't matter), then setChildren is called again on the tree, then the handler executes. It gets the focused element and calls setSelection. It fails because that exact object is not in the model.

This shows that when the Trait updates is elements in `onDidModelSplice`, it is not replacing the actual object in `nodes`.

![image](https://user-images.githubusercontent.com/323878/52361504-44d34900-29f3-11e9-9a37-5197fada71e4.png)

**Comment by roblourens** [comment]
Here are much simpler repro steps, I'm pretty sure it's the same issue.

- Do a search
- Select a search result to open the file
- Modify the file, e.g. insert a line after the search result
- cmd+shift+f to focus the search view
- cmd+down to focus that element
- cmd+down again to attempt to open the element
- Get 'tree element not found'

**Comment by joaomoreno** [comment]
@roblourens There's been quite a bit of refactoring around this, so I can't repro the last steps.

Basically here's how I intend it to be:

- Element identity is **always** object identity
- Certain tree node traits (focus, selection, collapse state) can be preserved across model changes, using their string identities. String identities are useful only to preserve view state across model changes.

Does that make sense?

**Comment by roblourens** [comment]
I can still repro issue with the last set of steps. This implies that getFocus or getSelection is returning an old object. Are you saying that focus will be preserved using the string ID, but getFocus may be invalid after refreshing the tree? Or is there still an issue for the tree?

But what you say makes sense. Do you think I should change the logic that refreshes the model to mutate existing objects instead of creating new objects with the same IDs, or should I just be aware that getFocus may return an invalid object?

**Comment by joaomoreno** [comment]
> Are you saying that focus will be preserved using the string ID, but getFocus may be invalid after refreshing the tree?

> Do you think I should change the logic that refreshes the model to mutate existing objects instead of creating new objects with the same IDs, or should I just be aware that getFocus may return an invalid object?

Focus should be _transferred_ to the new object model. So, `getFocus()` should return the new object model.

I can repro now... steps have different keyboard shortcuts on Linux. Definitely a bug!

**Comment by joaomoreno** [comment]
Pushed fix & test! Let me know how it goes.

**Comment by roblourens** [comment]
Yes, works now. Thanks!

**Comment by joaomoreno** [comment]
@roblourens Sweet, feel free to mark as `verified`!
